File: samples.fs/SqlSet.fs
namespace Samples.FSharp

open System
open System.Data
open System.Data.Common
open System.IO
open DbExtensions
open Samples.FSharp.Northwind

type SqlSetSamples(conn : DbConnection, log : TextWriter) =
   
   let log = log
   let productSet = conn.Set<Product>(new SqlBuilder("SELECT * FROM Products"), log)

   member this.AreThereAnyProducts() =
      productSet.Any()

   member this.DoAllProductsHaveUnitPrice() =
      productSet.All("NOT UnitPrice IS NULL")

   member this.DoSomeProductsAreOutOfStock() =
      productSet.Any("UnitsInStock = 0")

   member this.HowManyProductsAreOutOfStock() =
      productSet.Count("UnitsInStock = 0")

   member this.FirstProduct() =
      productSet.First()

   member this.SecondProduct() =
      productSet.Skip(1).First()

   member this.FirstOutOfStockProduct() =
      productSet.First("UnitsInStock = 0")

   member this.Top5ProductsWithLowestStock() =
      
      productSet.Where("UnitsInStock > 0")
         .OrderBy("UnitsInStock")
         .Take(5)
         .Select((fun r -> (r.GetString(0), r.GetInt16(1))), "ProductName, UnitsInStock")
         .AsEnumerable();

   member this.NamesOfOutOfStockProducts() =
         
      productSet.Where("UnitsInStock = 0")
         .Select((fun r -> r.GetString(0)), "ProductName")
         .AsEnumerable()
   
   member this.GetSpecificProduct() =
      productSet.SingleOrDefault("ProductID = 5")

File: samples/VisualBasic/SqlSet.vb
Imports System
Imports DbExtensions
Imports Samples.VisualBasic.Northwind

Public Class SqlSetSamples

   ReadOnly products As SqlSet(Of Product)

   Sub New(ByVal db As Database)
      Me.products = db.From(Of Product)("Products")
   End Sub

   Function AreThereAnyProducts() As Boolean
      Return products.Any()
   End Function

   Function DoAllProductsHaveUnitPrice() As Boolean
      Return products.All("NOT UnitPrice IS NULL")
   End Function

   Function DoSomeProductsAreOutOfStock() As Boolean
      Return products.Any("UnitsInStock = 0")
   End Function

   Function HowManyProductsAreOutOfStock() As Integer
      Return products.Count("UnitsInStock = 0")
   End Function

   Function FirstProduct() As Product
      Return products.First()
   End Function

   Function SecondProduct() As Product
      Return products.Skip(1).First()
   End Function

   Function FirstOutOfStockProduct() As Product
      Return products.First("UnitsInStock = 0")
   End Function

   Function Top5ProductsWithLowestStock() As IEnumerable

      Return products _
         .Where("UnitsInStock > 0") _
         .OrderBy("UnitsInStock") _
         .Take(5) _
         .Select(Function(r) New With {.Name = r.GetString(0), .UnitsInStock = r.GetInt16(1)}, "ProductName, UnitsInStock") _
         .AsEnumerable()

   End Function

   Function NamesOfOutOfStockProducts() As IEnumerable(Of String)

      Return products _
         .Where("UnitsInStock = 0") _
         .Select(Function(r) r.GetString(0), "ProductName") _
         .AsEnumerable()

   End Function

   Function GetSpecificProduct() As Product
      Return products.SingleOrDefault("ProductID = 5")
   End Function

End Class
